### DISCUSSION TITLE
Printing full traceback for exceptions

### DISCUSSION THREAD

**Discussion by scottstanie** [insarlab/MintPy#1017] [Polls]
Right now the CLI catches all exceptions and prints only the one top message without context:

https://github.com/insarlab/MintPy/blob/c305e13d929a0efba8b6ad418b2f48090ec2c2c2/src/mintpy/__main__.py#L627-L631

This means that for someone new (like me), I can't really figure out the error without a bunch of searching through the code for the matching phrase instead of having the full line number printed out in the traceback.

Although the stacktrace is printed with `log.debug("stacktrace:", exc_info=True)`, but it's unclear how to enable debug logging, as there's no `--debug` option.

**Comment by yunjunz** [comment]
Good point @scottstanie. It's essential to show the full traceback msg, one of Python's best parts. Is there a solution for it? Tagging @avalentino for ideas.

I mainly use individual scripts, such as `view.py`, and have not used the sub-command style `mintpy view` much.

**Reply by avalentino** [reply]
Hi @yunjunz  and @scottstanie.
As @scottstanie said the obvious solution is to add a "--debug" option to the CLI arguments,

```python
    parser.add_argument(
        "-d",
        "--debug",
        dest="loglevel",
        action="store_const",
        const="DEBUG",
        help="print debug messages",
    )
```

 and then set the logging level accordingly:

```python
logging.getLogger().setLevel(args.loglevel)
```

In case you need the full control over the logging level you could add something like the following:

```python
def _add_logging_control_args(parser, default_loglevel=DEFAULT_LOGLEVEL):
    """Add command line options for logging control."""
    loglevels = [logging.getLevelName(level) for level in range(10, 60, 10)]

    parser.add_argument(
        "--loglevel",
        default=default_loglevel,
        choices=loglevels,
        help="logging level (default: %(default)s)",
    )
    parser.add_argument(
        "-q",
        "--quiet",
        dest="loglevel",
        action="store_const",
        const="ERROR",
        help="suppress standard output messages, "
        "only errors are printed to screen",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        action="store_const",
        const="INFO",
        help="print verbose output messages",
    )
    parser.add_argument(
        "-d",
        "--debug",
        dest="loglevel",
        action="store_const",
        const="DEBUG",
        help="print debug messages",
    )
```

I typically try to organize my CLI programs according to the template that you can find in the following gist:
https://gist.github.com/avalentino/da77a67147403c8c25260ec7fd72d879

Another option, of course, is to change the log level for the traceback:
```python
log.exception("stacktrace:", exc_info=True)
```

... but in this case the traceback would be always logged in case of exception.

**Reply by yunjunz** [reply]
I like your way of full control on the logging level, and the traceback to be always logged in as well in case of exception, will follow a PR for it. Thank you @avalentino!

Update: here is the PR (https://github.com/insarlab/MintPy/pull/1020).
